Improvements for Reactive REST client (#59)
quarkus-qe/quarkus-test-plans: QUARKUS-1415.md

# QUARKUS-1415 Reactive REST Client improvements

JIRA link: https://issues.redhat.com/browse/QUARKUS-1415

> Test plan covers RESTEasy Reactive functionality provided by `quarkus-resteasy-reactive` and `quarkus-rest-client-reactive` extensions,

Reviewer: [QUARKUS-1415](https://issues.redhat.com/browse/QUARKUS-1415) is about Reactive REST Client improvements

with emphasis on multipart form data support.

> https://quarkus.io/version/main/guides/resteasy-reactive
> 
> https://quarkus.io/version/main/guides/rest-client-reactive

Reviewer: ```suggestion
Documentation:
 - https://quarkus.io/version/main/guides/resteasy-reactive
 - https://quarkus.io/version/main/guides/rest-client-reactive
```

## Related test plans

> - QUARKUS-1075

Reviewer: Make it clickable / link

> ## Scope of the testing

Reviewer: These topics are not covered:
 - support for ParamConverter
 - recursive sub-resources support, that is sub-sub-resources and so on work too,
 - UI / dev experience - List of clients in Dev UI and Links to client interfaces (if in user's app)

https://resteasy.dev/2021/04/05/blog-ParamConverter-with-Quarkus/ can be used for ParamConverter coverage

sub-resources can be a bit tricky topic, we should make sure one can't end up with infinite loop of sub-resources

UI/dev experience should be checked as part of the initial experiments, not necessary to de TD for it, but should be checked at least manually though

> - Upload and download of large files

Reviewer: `large file` is what exactly, please define concrete sizes?

> - Support for proxy authentication

Reviewer: How do you plan to simulate proxy?

Reviewer: Any concrete proxy? If so please include its name / image tag here

Author: Nothing in particular. Any recommendations?

Reviewer: No, you are preparing the plan, so I was wondering if you have something.

> - New configuration schema(https://github.com/quarkusio/quarkus/pull/17220)

Reviewer: ```suggestion
- New configuration schema introduced in https://github.com/quarkusio/quarkus/pull/17220
```

## Impact on test suites and testing automation:

> - Four new scenarios in QE TS, approximate time increase should be around several minutes, mostly due to uploading and downloading of big files 

Reviewer: In native the impact will be bigger, my guess is 4x ~4 minutes

# QUARKUS-1415 Reactive REST Client improvements

JIRA link: https://issues.redhat.com/browse/QUARKUS-1415

Test plan covers Reactive REST Client functionality provided by `quarkus-rest-client-reactive` extension.

Documentation:
https://quarkus.io/version/main/guides/resteasy-reactive
https://quarkus.io/version/main/guides/rest-client-reactive

## Related test plans
- [QUARKUS-1075](QUARKUS-1075.md)

## Scope of the testing

> - Upload and download of large(more than 2Gb) files, see https://github.com/quarkusio/quarkus/pull/22229 for details

Reviewer: Is it realistic to fetch 2 GB of data using REST Client?

Author: It is, but it will take some time(between 10 and 30 minutes, I guess)

Reviewer: ```suggestion
- Upload and download of large (more than 2GB) files, see https://github.com/quarkusio/quarkus/pull/22229 for details
```

let's see if native will be able to handle 2GB+ size :)

> - Upload and download of large(more than 2Gb) files, see https://github.com/quarkusio/quarkus/pull/22229 for details

Reviewer: ```suggestion
- Upload and download of large (more than 250MB) files, see https://github.com/quarkusio/quarkus/pull/22229 for details
```

Author: The original issue[1] explicitly mentions "files larger, than 2G". File of size 250MB will not really validate it.

[1] https://github.com/quarkusio/quarkus/issues/22218

- Support for proxy authentication
- New configuration schema introduced in https://github.com/quarkusio/quarkus/pull/17220
- Recursive sub-resources (see https://github.com/quarkusio/quarkus/pull/22417)
- Support for ParamConverter (related issue: https://github.com/quarkusio/quarkus/issues/23486)
- Manual check of Dev UI 

## Impact on test suites and testing automation:

> - Four new scenarios in QE TS, approximate time increase should be around several minutes, mostly due to uploading and downloading of big files. Additional impact for native is expected to be around 4 minutes 

Reviewer: ```suggestion
- Four new scenarios in QE TS
- Upload and download of large (more than 2GB) files will add several minutes to the execution time
- Additional impact for Native mode is expected to be around 4 minutes per new scenario
- Overall time execution increase is ~20 minutes
```
